File: scheduling/scheduling-annotation-processor/src/main/java/ru/tinkoff/kora/scheduling/annotation/processor/QuartzSchedulingGenerator.java
package ru.tinkoff.kora.scheduling.annotation.processor;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.*;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import java.util.Date;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;

public class QuartzSchedulingGenerator {
    public static ClassName scheduleWithTrigger = ClassName.get("ru.tinkoff.kora.scheduling.quartz", "ScheduleWithTrigger");
    public static ClassName scheduleWithCron = ClassName.get("ru.tinkoff.kora.scheduling.quartz", "ScheduleWithCron");

    private static final ClassName koraQuartzJobClassName = ClassName.get("ru.tinkoff.kora.scheduling.quartz", "KoraQuartzJob");
    private static final ClassName schedulingTelemetryClassName = ClassName.get("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetry");
    private static final ClassName schedulingTelemetryFactoryClassName = ClassName.get("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetryFactory");
    private static final ClassName triggerClassName = ClassName.get("org.quartz", "Trigger");
    private static final ClassName schedulerClassName = ClassName.get("org.quartz", "Scheduler");
    private static final ClassName triggerBuilderClassName = ClassName.get("org.quartz", "TriggerBuilder");
    private static final ClassName cronScheduleBuilderClassName = ClassName.get("org.quartz", "CronScheduleBuilder");
    private final Elements elements;
    private final ProcessingEnvironment processingEnv;

    public QuartzSchedulingGenerator(ProcessingEnvironment processingEnv) {
        this.elements = processingEnv.getElementUtils();
        this.processingEnv = processingEnv;
    }

    public void generate(TypeElement type, ExecutableElement method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var jobClassName = this.generateJobClass(type, method);
        var typeMirror = type.asType();

        var component = MethodSpec.methodBuilder(type.getSimpleName() + "_" + method.getSimpleName() + "_Job")
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .returns(jobClassName)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(TypeName.get(typeMirror), "object");


        var annotationType = ClassName.get((TypeElement) trigger.triggerAnnotation().getAnnotationType().asElement());
        if (annotationType.equals(scheduleWithTrigger)) {
            var tag = AnnotationUtils.<AnnotationMirror>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "value");
            var triggerParameter = ParameterSpec.builder(triggerClassName, "trigger").addAnnotation(AnnotationSpec.get(tag)).build();
            component.addParameter(triggerParameter);
            component.addCode("var telemetry = telemetryFactory.get(null, $T.class, $S);\n", typeMirror, method.getSimpleName().toString());
        } else if (annotationType.equals(scheduleWithCron)) {
            var identity = Optional.ofNullable(AnnotationUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "identity"))
                .filter(Predicate.not(String::isBlank))
                .orElse(type.getQualifiedName() + "#" + method.getSimpleName());
            var cron = AnnotationUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "value");
            var cronSchedule = CodeBlock.of("$S", cron);
            var configPath = AnnotationUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "config");
            if (configPath != null && !configPath.isBlank()) {
                var configClassName = this.generateCronConfigRecord(type, method, cron);
                var b = MethodSpec.methodBuilder(configClassName.simpleName())
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .returns(configClassName)
                    .addParameter(CommonClassNames.config, "config")
                    .addParameter(ParameterizedTypeName.get(CommonClassNames.configValueExtractor, configClassName), "extractor");

                if (cron != null && !cron.isBlank()) {
                    b.addStatement("var value = config.get($S)", configPath);
                    b.beginControlFlow("if (value instanceof $T.NullValue)", CommonClassNames.configValue)
                        .addCode("return extractor.extract($>\n")
                        .addCode("new $T.ObjectValue(value.origin(), $T.of($S, new $T.StringValue(value.origin(), $S)))", CommonClassNames.configValue, Map.class, "cron", CommonClassNames.configValue, cron)
                        .addCode("$<\n);\n")
                        .endControlFlow();
                } else {
                    b.addStatement("var value = config.get($S)", configPath);
                }
                b.beginControlFlow("if (value instanceof $T.StringValue str)", CommonClassNames.configValue)
                    .addStatement("var cron = str.value()")
                    .addCode("return extractor.extract($>\n")
                    .addCode("new $T.ObjectValue(value.origin(), $T.of($S, new $T.StringValue(value.origin(), cron)))", CommonClassNames.configValue, Map.class, "cron", CommonClassNames.configValue)
                    .addCode("$<\n);\n")
                    .nextControlFlow("else if (value instanceof $T.ObjectValue obj)", CommonClassNames.configValue)
                    .addStatement("return extractor.extract(obj)")
                    .nextControlFlow("else")
                    .addStatement("throw ru.tinkoff.kora.config.common.extractor.ConfigValueExtractionException.unexpectedValueType(value, $T.StringValue.class)", CommonClassNames.configValue)
                    .endControlFlow();

                module.addMethod(b.build());
                component.addParameter(configClassName, "config");
                cronSchedule = CodeBlock.of("config.cron()");
            } else {
                if (cron == null || cron.isBlank()) {
                    throw new ProcessingErrorException("Either value() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
                }
            }
            component.addCode("""
                var trigger = $T.newTrigger()
                  .withIdentity($S)
                  .withSchedule($T.cronSchedule($L))
                  .startAt(new $T(0))
                  .build();
                """.stripIndent(), triggerBuilderClassName, identity, cronScheduleBuilderClassName, cronSchedule.toString(), Date.class);
            if (configPath != null && !configPath.isBlank()) {
                component.addCode("var telemetry = telemetryFactory.get(config.telemetry(), $T.class, $S);\n", typeMirror, method.getSimpleName().toString());
            } else {
                component.addCode("var telemetry = telemetryFactory.get(null, $T.class, $S);\n", typeMirror, method.getSimpleName().toString());
            }
        } else {
            // never gonna happen
            throw new IllegalStateException();
        }

        component
            .addCode("return new $T(telemetry, object, trigger);\n", jobClassName);

        module.addMethod(component.build());
    }

    private ClassName generateCronConfigRecord(TypeElement type, ExecutableElement method, String defaultCron) {
        var configRecordName = NameUtils.getOuterClassesAsPrefix(method) + "CronConfig";
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();

        var config = TypeSpec.interfaceBuilder(configRecordName)
            .addOriginatingElement(method)
            .addAnnotation(AnnotationUtils.generated(JdkSchedulingGenerator.class))
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(CommonClassNames.configValueExtractorAnnotation)
            .addMethod(MethodSpec.methodBuilder("telemetry")
                .returns(CommonClassNames.telemetryConfig)
                .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                .build()
            );

        if (defaultCron != null && !defaultCron.isBlank()) {
            config.addMethod(MethodSpec.methodBuilder("cron")
                .returns(String.class)
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .addStatement("return $S", defaultCron)
                .build()
            );
        } else {
            config.addMethod(MethodSpec.methodBuilder("cron")
                .returns(String.class)
                .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                .build()
            );
        }
        CommonUtils.safeWriteTo(this.processingEnv, JavaFile.builder(packageName, config.build()).build());

        return ClassName.get(packageName, configRecordName);
    }


    private ClassName generateJobClass(TypeElement type, ExecutableElement method) {
        var className = NameUtils.generatedType(type, method.getSimpleName() + "_Job");
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var typeMirror = type.asType();
        var callJob = method.getParameters().isEmpty()
            ? CodeBlock.of("ctx -> object.$L()", method.getSimpleName())
            : CodeBlock.of("object::$L", method.getSimpleName());

        var typeSpec = TypeSpec.classBuilder(className)
            .addOriginatingElement(method)
            .addAnnotation(AnnotationUtils.generated(QuartzSchedulingGenerator.class))
            .superclass(koraQuartzJobClassName)
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addField(TypeName.get(typeMirror), "object", Modifier.PRIVATE, Modifier.FINAL)
            .addMethod(MethodSpec.constructorBuilder()
                .addModifiers(Modifier.PUBLIC)
                .addParameter(schedulingTelemetryClassName, "telemetry")
                .addParameter(TypeName.get(typeMirror), "object")
                .addParameter(triggerClassName, "trigger")
                .addCode("super(telemetry, $L, trigger);\n", callJob)
                .addCode("this.object = object;\n")
                .build())
            .build();

        var javaFile = JavaFile.builder(packageName, typeSpec).build();
        CommonUtils.safeWriteTo(this.processingEnv, javaFile);

        return ClassName.get(packageName, className);
    }
}

File: scheduling/scheduling-ksp/src/main/kotlin/ru/tinkoff/kora/scheduling/ksp/QuartzSchedulingGenerator.kt
package ru.tinkoff.kora.scheduling.ksp

import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSAnnotation
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.writeTo
import ru.tinkoff.kora.common.Tag
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.writeTagValue
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix
import java.util.Date

class QuartzSchedulingGenerator(val env: SymbolProcessorEnvironment) {
    private val koraQuartzJobClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.quartz", "KoraQuartzJob")
    private val schedulingTelemetryClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetry")
    private val schedulingTelemetryFactoryClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetryFactory")
    private val triggerClassName: ClassName = ClassName("org.quartz", "Trigger")
    private val schedulerClassName: ClassName = ClassName("org.quartz", "Scheduler")
    private val triggerBuilderClassName: ClassName = ClassName("org.quartz", "TriggerBuilder")
    private val cronScheduleBuilderClassName: ClassName = ClassName("org.quartz", "CronScheduleBuilder")


    fun generate(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val jobClassName = generateJobClass(type, function)
        val typeClassName = type.toClassName()
        val component = FunSpec.builder("_" + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job")
            .returns(jobClassName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("target", typeClassName);

        when (trigger.annotation.shortName.getShortName()) {
            "ScheduleWithTrigger" -> {
                val tag = trigger.annotation.findValue<KSAnnotation>("value")!!
                val tagAnnotationSpec = AnnotationSpec.builder(Tag::class)
                    .addMember(tag.findValue<List<KSType>>("value")!!.writeTagValue("value"))
                    .build()

                val triggerParameter = ParameterSpec.builder("trigger", triggerClassName)
                    .addAnnotation(tagAnnotationSpec)
                    .build()
                component.addParameter(triggerParameter)
                component.addCode("val telemetry = telemetryFactory.get(null, %T::class.java, %S);\n", typeClassName, function.simpleName.getShortName())
            }

            "ScheduleWithCron" -> {
                val identity = trigger.annotation.findValue<String>("identity").let {
                    if (it.isNullOrBlank()) {
                        type.qualifiedName!!.asString() + "#" + function.simpleName.getShortName()
                    } else {
                        it
                    }
                }
                val cron = trigger.annotation.findValue<String>("value") ?: ""
                var cronSchedule = CodeBlock.of("%S", cron)
                val configPath = trigger.annotation.findValue<String>("config")
                if (!configPath.isNullOrBlank()) {
                    val configClassName = this.generateCronConfigRecord(type, function, cron)
                    val b = FunSpec.builder(configClassName.simpleName)
                        .returns(configClassName)
                        .addParameter("config", CommonClassNames.config)
                        .addParameter("extractor", CommonClassNames.configValueExtractor.parameterizedBy(configClassName))
                    if (cron.isNotBlank()) {
                        b.addStatement("val value = config.get(%S)", configPath)
                        b.controlFlow("if (value is %T.NullValue)", CommonClassNames.configValue) {
                            addCode("return extractor.extract(\n")
                            addCode("  %T.ObjectValue(value.origin(), mapOf(%S to %T.StringValue(value.origin(), %S)))\n", CommonClassNames.configValue, "cron", CommonClassNames.configValue, cron)
                            addCode(")!!\n")
                        }
                    } else {
                        b.addStatement("val value = config.get(%S)!!", configPath)
                    }
                    b.controlFlow("if (value is %T.ObjectValue)", CommonClassNames.configValue) {
                        addStatement("return extractor.extract(value)!!")
                    }
                    b.controlFlow("if (value is %T.StringValue)", CommonClassNames.configValue) {
                        addCode("return extractor.extract(\n")
                        addCode("  %T.ObjectValue(value.origin(), mapOf(%S to %T.StringValue(value.origin(), value.value())))\n", CommonClassNames.configValue, "cron", CommonClassNames.configValue)
                        addCode(")!!\n")
                        nextControlFlow("else")
                        addStatement(
                            "throw %T.unexpectedValueType(value, %T.StringValue::class.java)",
                            ClassName("ru.tinkoff.kora.config.common.extractor", "ConfigValueExtractionException"),
                            CommonClassNames.configValue
                        )
                    }

                    builder.addFunction(b.build())
                    component.addParameter("config", configClassName);
                    cronSchedule = CodeBlock.of("config.cron()")
                }
                component.addCode(
                    """
                val trigger = %T.newTrigger()
                  .withIdentity(%S)
                  .withSchedule(%T.cronSchedule(%L))
                  .startAt(%T(0))
                  .build();
                """.trimIndent() + "\n", triggerBuilderClassName, identity, cronScheduleBuilderClassName, cronSchedule.toString(), Date::class.asClassName()
                );
                if (!configPath.isNullOrBlank()) {
                    component.addCode("val telemetry = telemetryFactory.get(config.telemetry(), %T::class.java, %S);\n", typeClassName, function.simpleName.getShortName())
                } else {
                    component.addCode("val telemetry = telemetryFactory.get(null, %T::class.java, %S);\n", typeClassName, function.simpleName.getShortName())
                }
            }
        }

        component
            .addCode("return %T(telemetry, target, trigger);\n", jobClassName)

        builder.addFunction(component.build())
    }


    private fun generateJobClass(type: KSClassDeclaration, method: KSFunctionDeclaration): ClassName {
        val className: String = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + method.simpleName.getShortName() + "_Job"
        val packageName: String = type.packageName.asString()
        val callJob = if (method.parameters.none { !it.hasDefault }) {
            CodeBlock.of("{ ctx -> target.%L() }", method.simpleName.getShortName())
        } else {
            CodeBlock.of("{ ctx -> target.%L(ctx) }", method.simpleName.getShortName())
        }
        val typeClassName = type.toClassName()
        val typeSpec: TypeSpec = TypeSpec.classBuilder(className)
            .generated(QuartzSchedulingGenerator::class)
            .superclass(koraQuartzJobClassName)
            .addSuperclassConstructorParameter(CodeBlock.of("telemetry"))
            .addSuperclassConstructorParameter(callJob)
            .addSuperclassConstructorParameter(CodeBlock.of("trigger"))
            .addProperty(
                PropertySpec.builder("target", typeClassName, KModifier.PRIVATE, KModifier.FINAL)
                    .initializer("target")
                    .build()
            )
            .primaryConstructor(
                FunSpec.constructorBuilder()
                    .addParameter("telemetry", schedulingTelemetryClassName)
                    .addParameter("target", typeClassName)
                    .addParameter("trigger", triggerClassName)
                    .build()
            )
            .build()
        FileSpec.get(packageName, typeSpec).writeTo(env.codeGenerator, false, listOf(type.containingFile!!))
        return ClassName(packageName, className)
    }

    private fun generateCronConfigRecord(type: KSClassDeclaration, function: KSFunctionDeclaration, defaultCron: String): ClassName {
        val configClassName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_CronConfig"
        val configType = TypeSpec.interfaceBuilder(configClassName)
            .addAnnotation(CommonClassNames.configValueExtractorAnnotation)
            .generated(QuartzSchedulingGenerator::class)
            .addFunction(FunSpec.builder("telemetry").returns(ClassName("ru.tinkoff.kora.telemetry.common", "TelemetryConfig")).addModifiers(KModifier.ABSTRACT).build())

        if (defaultCron.isBlank()) {
            configType.addFunction(FunSpec.builder("cron").addModifiers(KModifier.ABSTRACT).returns(STRING).build())
        } else {
            configType.addFunction(FunSpec.builder("cron").returns(STRING).addStatement("return %S", defaultCron).build())
        }

        FileSpec.get(type.packageName.asString(), configType.build()).writeTo(env.codeGenerator, false, listOf(type.containingFile!!))
        return ClassName(type.packageName.asString(), configClassName)

    }
}
